[CK_BUILDER] convolution testing

Snektron opened the pull request:
## Proposed Changes

This pull request adds the setup for a CK-Builder based device operation test framework. It supports multiple goals:
- A unified test framework for testing CK and CK Tile device operations.
- Dogfooding of CK-Builder by creating a sort of internal consumer of the API. This allows us to think about the API from an outside perspective, find missing or weird bits, etc.
- Take the opportunity to design a test system for a GPU-first world, where input generation, reference computations, and validation are all performed on the GPU.

## Design

A device operation test follows a Given-When-Then pattern. In short:
- Given: An operation and inputs
  - A signature and algorithm to define the operation.
  - A set of runtime arguments (tensor dimensions, filter dimensions, etc)
  - A set of input tensors with data
- When: Run the device operation
  - Invoke the device operation using the runtime arguments and input data, to produce output data.
- Then: Validate the outputs
  - Invoke a reference operation (on the GPU).
  - Compare the results.

The design of the test framework is described in detail in `experimental/builder/include/ck_tile/builder/testing/README.md`. Please see that file, so that I don't need to include it in the PR description.

## Implementation

This initial pull request adds an initial version of the following components:
- Runtime-argument definition.
- Passing/allocating inputs and outputs data.
- Invoking (CK) kernels
Not yet implemented:
- Input initialization
- Reference kernel invocation
- Output validation

The general implementation is described in `testing.hpp`. A small summary:

- `ckt::Args<SIGNATURE>` implements the run-time arguments structure for a particular operation defined by the `SIGNATURE`.
- Input and output tensors are passed around using `ckt::Inputs<SIGNATURE>` and `ckt::Outputs<SIGNATURE>`. This is a non-owning structure.
- `ckt::UniqueInputs<SIGNATURE>` and `ckt::UniqueOutputs<SIGNATURE>` are owning variants of the previous two structures. These can be allocated using an instance of the `Args` struct with the `alloc_inputs` and `alloc_outputs`.
- `ckt::run(instance, args, inputs, outputs)` invokes the `instance` with inputs `args` and `inputs`, and fills `outputs`.

Some additional details:
- All structures and functions are templated by `SIGNATURE`. The intention here is to use [dependent typing](https://en.wikipedia.org/wiki/Dependent_type) to allow all tests to use the same set of types and functions. As such, the contents of each structure and function may depend on the `SIGNATURE`, which is handled by specializing the structure for a particular set of `SIGNATURE`s using concepts.
- The goal is to implement the reference kernel using a specific type of `ALGORITHM`. This means that all structures and functions should gain a specialization for the reference implementation. Each test would include two invocations to `ckt::run()`, one for the device operation to-be-tested, and one for the reference implementation.
- For this reason, inputs and outputs are split up into separate structures. For each test, we allocate and initialize one set of inputs, and allocate two sets of outputs (one for the reference computation and one for the to-be-tested device operation)
- There are a few TODOs still in the code. These require refactors, rethinks, etc. I will create (internal) follow-up issues for most of these so that we can work on them separately and later.

shumway committed:
Add README.md for testing

shumway committed:
Add tensor_memory_manager.

Snektron committed:
ck-builder: tensor memory manager rebase fixes

This fixes some issues caused by the API being changed recently.
Also, this streamlines the ckt namespace to always be ck_tile::builder::test,
as this is already being used by other tests

Really, this commit should be squashed into the previous,
but I'm keeping it separate for brevity.

Snektron committed:
ck-builder: test arguments initial prototype

Snektron committed:
ck-builder: test system initial prototype

shumway commented on experimental/builder/test/conv/test_ckb_conv_fwd_1d_bf16.cpp:
```diff
@@ -36,4 +39,50 @@ TEST(FwdConvInstances,
+        .input_right_pad = {.width = 0},
+    };
+
+    (void)args;
```
Just use [[maybe_unused]] on the variable definition. It expresses intent more clearly and is easer to cleanup and maintain.

shumway commented on experimental/builder/include/ck_tile/builder/testing/type_traits.hpp:
```diff
@@ -0,0 +1,57 @@
+// Copyright (C) Advanced Micro Devices, Inc. All rights reserved.
```
Nit: Please stick the CK convention Avril established. We finally got aligned with the rest of the code (lowercase c, "or its affiliates").

// Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
// SPDX-License-Identifier: MIT

shumway commented on experimental/builder/test/conv/test_ckb_conv_fwd_2d_dl_fp16.cpp:
```diff
@@ -54,4 +59,104 @@ TEST(FwdConvInstances,
+        .input_right_pad = {.width = 0, .height = 0},
+    };
+
+    auto tmm = ckt::TensorMemoryManager<signature>(args);
```
I think I got this incorrect in my design notes. Construction doesn't need  args, just the allocation method.

shumway replied:
Also, it's clearer just to use:

ckt::TensorMemoryManager<kSignature> tmm;

We only need auto if we're getting a value from a factory.

shumway commented on experimental/builder/test/conv/test_ckb_conv_fwd_2d_dl_fp16.cpp:
```diff
@@ -54,4 +59,104 @@ TEST(FwdConvInstances,
+            .with_dl_thread_cluster(DlThreadCluster_8x2)
+            .with_dl_transfer(DlFwdTransfer);
+
+    auto args = ckt::ConvArgs<signature>{
```
Isn't our style kSignature for constants?

shumway commented on experimental/builder/test/conv/test_ckb_conv_fwd_2d_dl_fp16.cpp:
```diff
@@ -54,4 +59,104 @@ TEST(FwdConvInstances,
+
+    auto invoker = conv.MakeInvoker();
+
+    const auto input_desc  = tmm.input_descriptor;
```
We need to discuss this a lot more. All the input and output tensor info should come from the kSignature and the args.

The tensor memory allocator should be much simpler:

```
ckt::TensorMemoryManager<kSignature> tensor_memory;
tensor_memory.allocate(args);
````

Snektron replied:
> We need to discuss this a lot more. All the input and output tensor info should come from the kSignature and the args.

That was my first thought as well, but the simplest implementation reuses CK's `ConvParam` implementation to do the allocation and this avoids reconstructing that type. Its not a permanent location, though. That type is also needed for the actual invocation. I'm thinking about another structure that converts the `ConvArgs` into all of this extra information, which is then passed into the tmm or something.

shumway reviewed (commented):
This is a a great start. We should iterate some more on the tensor manager usage and design.

Snektron committed:
ck-builder: fix non-standardized copyright comments

Snektron committed:
ck-builder: new prototype

Snektron committed:
ck-builder: group testing inputs/outputs into a separate structure

This is basically the return of the tensor memory manager after all,
except that the design is more closely tied to the actual operation.
Using a struct allows us to add additional input/output tensors
without breaking code (by defaulting those new parameters). Note
that the tensors are split into a separate inputs/outputs because we
usually want to allocate the output _twice_: once for the real
computation and once for the reference computation.

Snektron committed:
ck-builder: simplify prototype naming; start docs

Snektron committed:
ck-builder: update testing readme

Snektron committed:
ck-builder: testing documentation

Snektron committed:
ck-builder: HipStatusMatcher

This matcher can be used to check HIP status codes and provide
nice and readable error messages.

Snektron committed:
ck-builder: tensor_buffer.hpp tests

Snektron committed:
ck-builder: conv_fwd.hpp tests

Snektron committed:
ck-builder: add example end-to-end test in conv fwd 2d fp16

amd-anclark commented on experimental/builder/include/ck_tile/builder/testing/README.md:
This reads very well is is straightforward. Very nice work!

amd-anclark commented on experimental/builder/include/ck_tile/builder/testing/README.md:
````diff
@@ -0,0 +1,329 @@
+EXPECT_THAT(validator.result(), validator.is_ok());
+```
+
+The `is_ok()` matcher checks that the output is numerically correct within acceptable tolerances. The `Validator` can also provide more detailed diagnostics, such as:
````
I don't see the custom matchers existing yet, but just recommending the name of the is_ok() matcher should convey what is actually being checked. Maybe something like matches_reference_output(), is_within_tolerance(), is_numerically_correct() for example.
```suggestion
The `matches_reference_output()` matcher checks that the output is numerically correct within acceptable tolerances. The `Validator` can also provide more detailed diagnostics, such as:
```

Snektron replied:
I think that the details of this will change indeed. Ive updated the docs to your suggestion for now.

amd-anclark commented on experimental/builder/include/ck_tile/builder/testing/README.md:
```diff
@@ -0,0 +1,329 @@
+
+#### Building the Reference Kernel
+
+The reference kernel is just another kernel instance of the builder, one that's been externally verified to produce the correct results. As this kernel is also running on the GPU, we can use it to perform tests far more quickly than when comparing the outputs to a CPU-based reference implementation.
```
I really like this approach!
I'm sure you've already thought of it this, but just in case not, we should be adding automated checks to verify the known reference kernels are consistent over time.

Snektron replied:
Yep, the reference kernel should be "known good" in the context of the CK Builder test infrastructure. I believe that the current implementation already has this.

shumway commented:
This needs a PR description.

shumway commented on experimental/builder/include/ck_tile/builder/testing/README.md:
```diff
@@ -0,0 +1,329 @@
+
+This directory contains testing utilities designed to simplify the process of writing unit tests for GPU kernels built with `ck_tile::builder`. These utilities enable a clean, expressive **Given-When-Then** (Given-When-Then) testing pattern that separates test setup, execution, and validation.
+
+## Overview
```
Refer back to CK builder documentation, see Ville's example in https://github.com/ROCm/composable_kernel/blob/develop/experimental/builder/include/ck_tile/builder/factory/README.md.

Snektron replied:
I've also updated the main file to link to the different modules.

shumway commented on experimental/builder/include/ck_tile/builder/testing/conv_fwd.hpp:
```diff
@@ -0,0 +1,259 @@
+    Ops::BElementwiseOp b_elementwise_op;
+    Ops::CDEElementwiseOp cde_elementwise_op;
+
+    // TODO: We shouldn't need to call into an internal namespace here.
```
This is an interesting problem. The internal namespace of the factory is private implementation translate the signature and algorithm to instantiate CK kernels. In principle it should not be used here, and this reuse is hinting that we may need some more general functionality for processing signature descriptions. I like your approach of doing the reuse and using a TODO, this helps us learn and keeps us from over-designing too early.

Snektron replied:
I'm more and more leaning to rewriting the functionality to turn the `Args` into a descriptor, for a few reasons, one of which is related to this. But that can be follow-up work.

Snektron replied:
On a related note, I'm as of yet not really sure what to do with the `Ops::` fields. We will either need to supply an abstraction here or map these to implementation-specific element-wise operation components.

shumway commented on experimental/builder/include/ck_tile/builder/testing/conv_fwd.hpp:
```diff
@@ -0,0 +1,259 @@
+        return TensorDescriptor<OUTPUT_TYPE>(desc.GetLengths(), desc.GetStrides());
+    }
+
+    /// Convert the Args structure into a CK conv_param structure. This
```
Very good, this logic is going to be useful. We'll have some choice how much of this we put into the args and how much we put in the run mechanism. My guess is this will eventually move out of the args and be part of our execution interface.

Snektron replied:
Yep, I agree. The main reason that its still here right now is so that its easier to create the descriptors (that will still need to be associated with the Args).

Actually, I think that eventually there will be headers that have a bunch of integration for CK / CK Tile / etc.

shumway commented on experimental/builder/include/ck_tile/builder/testing/conv_fwd.hpp:
```diff
@@ -0,0 +1,259 @@
+    ck::utils::conv::ConvParam to_ck_conv_param() const
+    {
+        const auto to_vector = [](const auto& extent) {
+            std::vector<ck::index_t> result;
```
This is much simpler:

switch (SPATIAL_DIM){
  case 1: return std::vector<ck::index_t> {extent.width};
  case 2:  return std::vector<ck::index_t> { extent.height, extent.width};
  default: return std::vector<ck::index_t> {extent.depth, extent.height, extent.width};
}

It's silly to optimize away the if-branch but use reserve() and push_back(). If you really want to keep the constexpr-if you can drop the switch-case, but I think the readability is worth a tiny cost in a lambda.

Snektron replied:
Actually because of the way that the struct is defined (with templates like this), the check has to be constexpr, otherwise there is an error about an inactive field.

Snektron replied:
Implemented your suggestion using `if constexpr`

shumway commented on experimental/builder/include/ck_tile/builder/testing/extent.hpp:
```diff
@@ -0,0 +1,40 @@
+    size_t depth  = 1;
+};
+
+using Extent1D = Extent<1>;
```
Is there a reason to use templates here? If we are accessing the dimensionality with metaprogramming, then use templates, otherwise just naming them  Extent1D, Extent2D, Extent3D is more clear (especially since the field names are hard-coded). Doing both with the alias is a little confusing.

Snektron replied:
The main thought behind this was to prevent users from accidentally writing `{.width = 1, .height = 1},` when the spatial dim is 1. It does make the conversion to vector slightly more complicated though. Do you think that its better to have a single 3D extent structure and ignore or assert those fields are 1 instead?

shumway commented on experimental/builder/include/ck_tile/builder/testing/tensor_buffer.hpp:
```diff
@@ -0,0 +1,212 @@
+    /// This function is invoked by `std::unique_ptr` when memory that the
+    /// pointer represents should be freed. In our implementation, we just
+    /// pass it directly to `hipFree`.
+    void operator()(std::byte* ptr) const
```
I need to lobby to change our style to make this a one liner!

shumway reviewed (commented):
This is great. It looks like you have to resolve merge conflicts, so I'll hold off on approval.

I think we'll improve on this design, the helpers, and concepts as we use this, so this looks like it's the right level of engineering. Your source comments may be more than we need, but thank you for taking the time, especially as we are trying to align everyone on a design.

Snektron committed:
ck-builder: simplify extent usage

Snektron committed:
ck-builder: update testing doc

spolifroni-amd commented on experimental/builder/include/ck_tile/builder/testing/README.md:
```diff
@@ -0,0 +1,347 @@
+1. **Given**: Set up the preconditions and test data
+2. **When**: Execute the action being tested
+3. **Then**: Verify the expected outcome
+
```
I'd move this up personally. Make it clear what it is from the get-go.

Snektron replied:
Hmm, I'm not sure that I like that. Right now the pattern and its details fits nicely into its own section, and moving the explanation of the given-then-when pattern up starts the "implementation details" of the readme to soon imo. I've left it like this for now.

spolifroni-amd commented on experimental/builder/include/ck_tile/builder/testing/README.md:
```diff
@@ -0,0 +1,347 @@
+
+#### Tensor Memory Management
+
+Tensor memory is passed around using the `Inputs<SIGNATURE>` and `Outputs<SIGNATURE>` structures. These group all inputs and outputs for an operation. Note that these structures do not "own" the memory inside: They only logically group the inputs so that they can be passed around as a common type. The amount of inputs and outputs may differ depending on the `SIGNATURE`, and this avoids having to  ...
```
```suggestion
Tensor memory is passed using the `Inputs<SIGNATURE>` and `Outputs<SIGNATURE>` structures. These group all inputs and outputs for an operation. Note that these structures do not "own" the memory inside: They only logically group the inputs so that they can be passed as a common type. The amount of inputs and outputs may differ depending on the `SIGNATURE`, and this avoids having to pass additional values and accept additional parameters in those situations.
```

"passed around" sounds like a tray of hors d'oeuvres at a party :)

It would be good to say what the memory is being passed to/from.

spolifroni-amd commented on experimental/builder/include/ck_tile/builder/testing/README.md:
````diff
@@ -0,0 +1,347 @@
+auto outputs = ck_tile::test::allocate_outputs(args);
+```
+
+Note that these functions merely _allocate_ memory: After allocation, the memory is yet uninitialized.
````
```suggestion
Note that these functions merely _allocate_ memory: After allocation, the memory is still uninitialized.
```

spolifroni-amd reviewed (changes requested):
some feedback re. the readme.

aosewski commented on experimental/builder/include/ck_tile/builder/testing/extent.hpp:
```diff
@@ -0,0 +1,36 @@
+namespace ck_tile::builder::test {
+
+/// This structure describes a 1-, 2-, or 3-D extent. Its used to
+/// communicate 1-, 2- or 3-D sizes and strides of tensors.
```
I don't see information about strides - or you assume that tensors are always packed?

Snektron replied:
Thats correct, though thats unrelated to this type (see its use in conv_fwd.hpp). The functionality from CK that I'm using to construct tensor descriptors does not actually support non-packed tensors right now, and rather than changing that in this PR I figured to do it as part of follow-up work. I've already created an internal ticket to keep track of that.

Snektron committed:
ck-builder: skip end to end test on non-gfx9

Snektron committed:
fix check_copyright_year interpreter

/bin/bash is not guaranteed to exist on Linux. Signed,
a NixOS user

Snektron committed:
ck-builder: reduce conv fwd testing size

This test allocated 24GB of memory, too much for 16GB cards.

Snektron committed:
ck-builder: fix copyrights
